Make some blocks conditional on relevant variables

whi-tw opened the pull request:
Some config blocks are only relevant if the user has certain variables set to a specific value.
If these blocks are present when they don't need to be, they can force recreation of the `google_container_cluster` in instances where it should not need to be recreated.

<dl>
  <dt>google_container_cluster.node_config</dt>
  <dd>This is only necessary if <code class="notranslate">var.confidential_nodes_enabled</code> is <code class="notranslate">true</code>.<br />If it is set, changing <code class="notranslate">var.machine_type</code> will trigger a recreation.</dd>
  <dt>google_container_cluster.private_cluster_config</dt>
  <dd>This has been identified by terraform a few times as the <code class="notranslate"># forces replacement</code> line. I'm not 100% sure why, but as soon as this block was made conditional, the replacement was no longer forced on some unrelated variable changes.</dd>
</dl>

Additionally, refer to `google_container_cluster.primary` from `google_container_node_pool.primary_preemptible_nodes` by `id`, not `name` to help terraform understand the relationship better (the individual commit message explains this better)

whi-tw committed:
Make the node_config block conditional

Rather than always having a node_config block, make it conditional on
`var.confidential_nodes_enabled` (as it is only used for those).

Without this change:
   - changing `var.machine_type` causes terraform to delete and recreate
     the whole cluster.
   - The cluster is marked as 'changed' every time `terraform plan` is
     run, even if nothing has changed.

whi-tw commented:
Also, this means it won't spin up the `standard` nodes on first deploy.

Neutrollized commented:
Hi!  Thanks for your PR!  At a glance it makes sense, but I'm on vacation right now, I'll take a closer look when I'm back.  

Cheers,
GY

whi-tw committed:
Make private_cluster_config block conditional

google_container_cluster.private_cluster_config is only relevant if
enable_private_nodes is true.

With this change, fewer unnecessary recreates are triggered when
changing other variables.

whi-tw committed:
Refer to cluster by id, not name

If the cluster is replaced, but the node pool does not 'need to be',
the node-pool is not flagged for replacement when it should be.

It is deleted within GKE when the cluster itself is removed, but
terraform's state does not reflect this. When terraform then creates
the 'new' cluster, no node pool is created as, as far as terraform is
aware, it already exists.

By referring to the cluster by id which is not known at plan time for
a creation, the node pool is correctly flagged for replacement.

ref: https://github.com/hashicorp/terraform-provider-google/issues/
   11164#issuecomment-1050189855

Neutrollized commented:
LGTM!  Thanks for the PR!
